docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,421 @@
+        "text": "I completely agree!",
+        "author": "<some base64>",
+        "signature": "<some base64>"
+    })
+})
+const change = Automerge.getChanges(doc, updatedDoc)
+const changeBytes = uint8ToBase64(change)
+```
+
+What we do here is load the automerge document from it's history, then use the
+automerge Javscript library to mutate the document (the `Automerge.change`
+call) and then finally get the change between the original version of the 
+document and the new one. 
+
+Now that we have the change we can make a `PATCH` request to 
+`<radicle-implementation>/projects/<project URN>/metadata/<metadata ID>` with
+the following contents:
+
+```json
+{
+    "changes": {
+        "type": "automerge",
+        "change": "<some base64>
+    }
+}
+```
+
+This endpoint will return an error if the change does not match the schema of
+the object. Otherwise the change will be merged in to the object and announced
+to the network.
+
+
+### Changing the schema
+
+It's a few months on, everyone is very happy with issues except for one thing,
+there is no way to react to a comment with an emojii. To accomodate this we
+modify the schema to add a `reaction` field to the `comment` schema. Now when
+we create an issue, as well as passing the schema, we also pass a schema 
+migration. Like so:
+
+```json
+{
+    "typename": ...,
+    "schema": ...,
+    "schema_migrations": [
+        {
+            "type": "add_field",
+            "path": "comments",
+            "name": "reaction",
+            "schema": {
+                "type": "string",
+                "maxLength": 1,
+            }
+        }
+    ],
+    "data": ...
+}
+```
+
+And when updating an object:
+
+```json
+{
+    "changes": ...,
+    "schema_migrations": <as above>
+}
+```
+
+There are restrictions on these migrations. You can only ever add optional 
+fields, or fields with a default value. This schema migration must be bundled
+with the application, must as database migrations are bundled in the source
+code of web 2.0 applications.
+
+## Message storage and transmission
+
+The property of CRDTs that we care about is strong eventual consistency: any
+two nodes which have received the same set of messages will merge to the same
+state. A major part of the design of collaborative objects therefore is how we
+store and transmit these messages. Messages in this context are just binary
+blobs which need to be passed to the CRDT implementation in any order.
+
+We add a new entry to the `signed_refs` of a project at
+`refs/namespaces/<project>/remotes/refs/remotes/<peer>/rad/collaborative-objects`.
+This points to a tree with the following structure:
+
+```
+.
+|--<object 1 ID>
+|  |--manifest
+|  |--schema
+|  |  |--<schema change 1>
+|  |  |--<schema change 2>
+|  |--<change 1>
+|  |--<change 2>
+|--<object 2 ID>
+|  ...
+```
+
+We refer to a directory in this tree as an "object directory" and the tuple 
+`(manifest, [schema changes], [changes])` as an "object state".
```

Comment:
So I'm not clear on what the storage is across peers. Taking the example of the link team collaborating we would have a the following structure in our monorepo:

```
├── namespaces
│   ├── hnrkc3expus8az3s5iehfyi5qegyh3841nwxo
│   │   └── refs
│   │       ├── rad
│   │       │   ├── id
│   │       │   ├── ids
│   │       │   │   ├── alex-person
│   │       │   │   ├── finto-person
│   │       │   │   ├── kim-person
│   │       │   │   └── xla-person
│   │       │   ├── self
│   │       │   └── signed_refs
│   │       └── remotes
│   │           ├── alex-peer
│   │           │   ├── heads
│   │           │   │   └── main
│   │           │   └── rad
│   │           │       ├── id
│   │           │       ├── ids
│   │           │       │   ├── alex-person
│   │           │       │   ├── finto-person
│   │           │       │   ├── kim-person
│   │           │       │   └── xla-person
│   │           │       └── signed_refs
│   │           ├── kim-peer
│   │           │   ├── heads
│   │           │   │   └── main
│   │           │   └── rad
│   │           │       ├── id
│   │           │       ├── ids
│   │           │       │   ├── alex-person
│   │           │       │   ├── finto-person
│   │           │       │   ├── kim-person
│   │           │       │   └── xla-person
│   │           │       └── signed_refs
│   │           └── xla-peer
│   │               ├── heads
│   │               │   └── main
│   │               └── rad
│   │                   ├── id
│   │                   ├── ids
│   │                   │   ├── alex-person
│   │                   │   ├── finto-person
│   │                   │   ├── kim-person
│   │                   │   └── xla-person
│   │                   ├── self
│   │                   └── signed_refs
```

Now we're adding a new reference `rad/collaborative_objects`. How would that be split between my top-level `rad` tree and `alex-peer`, `kim-peer`, and `xla-peer`? Would we maintain a consistent view of all the changes we're aware of or just our own changes? If the latter, do we store the computed document? In the current iteration, I don't see where that would fit in `(manifest, [schema changes], [changes])`.

Reply:
I've pushed a change which slightly modifies the layout, but this question is still basically the same. My current thinking is that the changes each remote is aware of stay in their respective remotes - we don't store the computed document in the repository (though we may want to cache it somewhere). When we modify an object then we copy all the changes that the modified object depends on into our view of that object and announce it.

Reply:
> we don't store the computed document in the repository (though we may want to cache it somewhere).

Ah, I was thinking automerge takes the whole document, but looked back and saw it loads the document from a set of changes. I suppose applications could decide to cache the computed document.

> I've pushed a change which slightly modifies the layout, but this question is still basically the same. My current thinking is that the changes each remote is aware of stay in their respective remotes

Something that's brewing in my head is what happens when there's a number of hops to something like a reply. Say you created an issue that an Ink-n-Switch collaborator replies to and their colleague replies to that. Which one of those replies would _I_ see and how would those changes reach me?

Reply:
I _think_ this is what I address in the "strange perspectives" section. It's effectively the same model as for receiving changes in the code in a repository (this is a deliberate choice, hopefully we only have to educate users about it once that way). The problems and solutions will likely be the same - things like seed nodes which listen for changes you might be "interested" in but which are not part of your tracking graph and notify you of them.